src/MailboxesParser/MailboxesParser.php

```diff
@@ -0,0 +1,245 @@
+        self::SENT      => 'Sent',
+        self::SPAM      => 'Spam',
+        self::TRASH     => 'Trash',
+        self::TEMPLATES => 'Templates',
+        self::ARCHIVES  => 'Archives',
+    ];
+
+    protected $specialFoldersOrder = [
+        self::INBOX     => 1,
+        self::SENT      => 2,
+        self::DRAFT     => 3,
+        self::TEMPLATES => 4,
+        self::ARCHIVES  => 10000,
+        self::SPAM      => 20000,
+        self::TRASH     => 30000,
+    ];
+
+    /**
+     * MailboxesTree constructor.
+     *
+     * @param MailboxInterface[] $mailboxes
+     */
+    public function __construct($mailboxes)
+    {
+        $this->mailboxes = $mailboxes;
+    }
+
+    public function setLanguage($lang)
+    {
+        $path = __DIR__ . DIRECTORY_SEPARATOR . 'languages' . DIRECTORY_SEPARATOR . $lang . DIRECTORY_SEPARATOR . 'names.php';
+        if (!is_file($path)) {
+            throw new  MailboxesParserException(\sprintf('File for language names %s does not exist', $path));
+        }
+        $names = require $path;
+        $this->setSpecialFoldersNames($names);
+
+        $path = __DIR__ . DIRECTORY_SEPARATOR . 'languages' . DIRECTORY_SEPARATOR . $lang . DIRECTORY_SEPARATOR . 'ids.php';
+        if (!is_file($path)) {
+            throw new  MailboxesParserException(\sprintf('File for language ids %s does not exist', $path));
+        }
+        $ids = require $path;
+        foreach ($ids AS $specialFolder => $idsArray) {
+            foreach ($idsArray AS $id) {
+                $this->addSpecialFolderId($specialFolder, $id);
+            }
+        }
+    }
+
+    protected function parse()
+    {
+        $this->folders = [];
+        usort($this->mailboxes, [$this, "sortByMailboxName"]);
+        foreach ($this->mailboxes AS $k => $mailbox) {
+            $mailboxName = $mailbox->getName();
+            $folder = new ParsedMailbox();
+            $folder->setMailbox($mailbox);
+            $folder->setMailboxName($mailboxName);
+            $special = $this->getSpecialFolder($mailboxName);
+            $folder->setSpecial($special);
+            $folder->setName($special ? $this->specialFoldersNames[$special] : $this->getName($mailboxName, $mailbox->getDelimiter()));
+            $folder->setOrder($special ? $this->specialFoldersOrder[$special] : $this->getOrder($mailboxName, $mailbox->getDelimiter()));
+            $folder->setLevel($this->getFolderLevel($mailboxName, $mailbox->getDelimiter()));
+            $folder->setDelimiter($mailbox->getDelimiter());
+            $this->folders[] = $folder;
+        }
+
+        usort($this->folders, [$this, "sortByOrder"]);
+
+        return $this->folders;
+    }
+
+
+    protected function sortByMailboxName(MailboxInterface $a, MailboxInterface $b)
+    {
+        return ($a->getName() < $b->getName()) ? -1 : 1;
+    }
+
+    protected function sortByOrder(ParsedMailbox $a, ParsedMailbox $b)
+    {
+        return ($a->getOrder() < $b->getOrder()) ? -1 : 1;
```

Comment:
We can use here the [Spaceship operator](https://wiki.php.net/rfc/combined-comparison-operator)